### ISSUE TITLE
HTTP/2 window size is not correctly updated for big files

### ISSUE THREAD

**Issue by sturcotte06** [h2o/h2o#1515]
When streaming a big file in HTTP/2, the on_read protocol callback is only invoked for the first 32k buffer. It is not rescheduled afterwards. The streaming handler is therefore invoked a couple of times, but never get called with is_end_stream=1.

The request ends up deadlocking, as further socket reads never occur.

#### h2o/http2/connection.c 
1. expect_preface
1. handle_settings_frame
1. handle_window_update_frame
1. handle_headers_frame
1. on_read n times until buffer is full
1. handle_data_frame n times until buffer is empty
1. handle_settings_frame with last couple of bytes
1. on_read is never called again

### Lib versions
h2o version: 2.2.3-DEV
uv version: 1.16.0

### Minimal code
[main.zip](https://github.com/h2o/h2o/files/1524051/main.zip)
### How to reproduce

```bash
gcc main.c -lpthread -lh2o -luv -lcrypto -lssl  -o main
./main &
dd if=/dev/zero of=test bs=1G count=1
curl -vvv --http2-prior-knowledge --data-binary @test -H"Content-Type: application/octet-stream" http://localhost:8080
```

**Comment by sturcotte06** [comment]
I'm including the trace output from nghttp.

[nghttp.txt](https://github.com/h2o/h2o/files/1524420/nghttp.txt)

**Comment by sturcotte06** [comment]
I've got an update on this issue.

In **connection,c**, update_input_window() is invoked in handle_data_frame() and h2o_http2_encode_frame_header() is invoked in handle_settings_frame() with a stream_id of 0 instead of with frame->stream_id.

In the current master version, a file of 65534 bytes (64K - 1) can be successfully uploaded. From 64K on onward, send transfer get stopped because of a wrong input window update.

With the change to lines 633 and 801 (changing stream_id parameter from 0 to frame->stream_id), it allows files up to 16MB to upload correctly. However, it still blocks the client when a file is bigger.

It seems like a bug with window size handling which prevent the client from sending any new data frame.

**Comment by objd** [comment]
The issue was that the window size needed to be updated for both the connection and the stream on data frame received.

I've proposed a fix: #1516
